pubky/src/native/internal/pkarr.rs

```diff
@@ -1,36 +1,140 @@
+                    Some(h) => h,
+                    // If there's an existing record but no embedded host, there's nothing to update.
+                    None => return Ok(()),
+                }
+            }
+
+            // No host was provided, and no existing record, so nothing to publish.
+            (None, None) => return Ok(()),
+        };
+
+        // Determine if we should publish based on the given strategy.
+        let should_publish = match strategy {
+            PublishStrategy::Force => true,
+            PublishStrategy::IfOlderThan => {
+                match existing {
+                    Some(ref record) => {
+                        let now_micros = Timestamp::now();
+                        let record_micros = record.timestamp().as_u64();
+                        now_micros - record_micros > Timestamp::from(self.max_record_age_micros)
+                    }
+                    None => true, // If there's no record yet, we publish.
                 }
             }
+        };
+
+        if should_publish {
+            self.publish_homeserver_inner(keypair, &host_str, existing)
+                .await?;
         }
 
-        let svcb = SVCB::new(0, host.try_into()?);
+        Ok(())
+    }
 
+    /// Internal helper that builds and publishes the PKarr record.
+    /// Uses an optionally pre-resolved record to avoid re-resolving.
+    async fn publish_homeserver_inner(
+        &self,
+        keypair: &Keypair,
+        host: &str,
+        existing: Option<SignedPacket>,
+    ) -> Result<()> {
+        let mut builder = SignedPacket::builder();
+        if let Some(ref record) = existing {
+            // Append any records (except those already starting with "_pubky") to our builder.
+            for answer in record.resource_records("_pubky") {
+                if !answer.name.to_string().starts_with("_pubky") {
+                    builder = builder.record(answer.to_owned());
+                }
+            }
+        }
+        let svcb = SVCB::new(0, host.try_into()?);
         let signed_packet = SignedPacket::builder()
             .https("_pubky".try_into().unwrap(), svcb, 60 * 60)
             .sign(keypair)?;
-
         self.pkarr
-            .publish(&signed_packet, existing.map(|s| s.timestamp()))
+            .publish(&signed_packet, existing.map(|r| r.timestamp()))
             .await?;
+        Ok(())
+    }
 
+    /// Helper to extract the current homeserver host from a signed PKarr record.
+    /// Iterates over the records with name "_pubky" and returns the first SVCB target found.
+    fn extract_host_from_record(record: &SignedPacket) -> Option<String> {
+        record
+            .resource_records("_pubky")
+            .find_map(|rr| match &rr.rdata {
+                RData::SVCB(svcb) => Some(svcb.target.to_string()),
+                RData::HTTPS(https) => Some(https.0.target.to_string()),
+                _ => None,
+            })
+    }
+}
```

Comment:
SignedPacket contains multiple records. So I suggest renaming the variable for correctness.

```suggestion
    /// Helper to extract the current homeserver host from a signed PKarr packet.
    /// Iterates over the records with name "_pubky" and returns the first SVCB target found.
    fn extract_host_from_packet(packet: &SignedPacket) -> Option<String> {
        packet
            .resource_records("_pubky")
            .find_map(|rr| match &rr.rdata {
                RData::SVCB(svcb) => Some(svcb.target.to_string()),
                RData::HTTPS(https) => Some(https.0.target.to_string()),
                _ => None,
            })
    }
}
```